lucetc/src/function.rs

```diff
@@ -16,10 +16,117 @@ use memoffset::offset_of;
 use std::collections::HashMap;
 use wasmparser::Operator;
 
+enum ScopeControlFlow {
+    // this scope has linear control flow, where `bool` is if it's still reachable
+    Linear(bool),
+    // this scope has conditional control flow, `bool` is for the branch being reachable,
+    // where `Option<bool>` describes reachability of the alternate branch
+    Conditional(bool, Option<bool>)
+}
+impl ScopeControlFlow {
+    pub fn terminates(&self) -> bool {
+        match self {
+            ScopeControlFlow::Linear(reachable) => { !reachable },
+            ScopeControlFlow::Conditional(_, None) => {
+                // this never terminates because the maybe-terminal branch can be skipped
+                false
+            },
+            ScopeControlFlow::Conditional(fallthrough, Some(alternate)) => {
+                // an if terminuates if the fallthrough terminates and the alternate, when present,
+                // terminates
+                !fallthrough && !alternate
+            }
+        }
+    }
+}
+
+struct ScopeTracker {
```

Comment:
would appreciate some guide-level documentation of this machinery, including an explanation of the interrelationship between  "terminates" and "reachable" in this setting